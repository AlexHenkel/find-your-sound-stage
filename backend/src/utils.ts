import { isAfter, isBefore, startOfDay } from "date-fns";
import { Studio } from "./types";
import { SearchInput } from "./inputs";
import { StudioData } from "./data/types";

type SearchString = keyof SearchInput & keyof Studio;
const stringKeys: SearchString[] = ["city", "state", "country"];

export const filterStages = (
  searchQuery: SearchInput | undefined,
  data: StudioData[]
) =>
  searchQuery
    ? data.filter((studio) =>
        // Either filter not present, or filter match with value
        stringKeys.every(
          (k) =>
            !searchQuery[k] ||
            // Use lower case to expand results
            studio[k].toLowerCase() === searchQuery[k]?.toLowerCase()
        )
      )
    : // If query not present, just return all data
      data;

export const filterDates = (
  searchQuery: SearchInput | undefined,
  filteredStudios: StudioData[]
) => {
  // Make sure we compare with the start of day, so `isBefore` or `isAfter` are false with same dates
  const queryStartDate = searchQuery?.startDate
    ? startOfDay(searchQuery.startDate)
    : undefined;
  const queryEndDate = searchQuery?.endDate
    ? startOfDay(searchQuery.endDate)
    : undefined;

  const response: Studio[] = [];

  // Search for matching dates on filtered studios
  filteredStudios.forEach((studio) => {
    studio.sound_stages.forEach((stage) => {
      // Use regular for, so we can break and stop searching for dates on same stage
      for (const date of stage.available_dates) {
        // Either filter not present, or filter date is same or after start date
        if (
          !queryStartDate ||
          !isBefore(queryStartDate, startOfDay(new Date(date.start_date)))
        ) {
          // Either filter not present, or infinite availability, or filter date is same or before end date
          if (
            !queryEndDate ||
            !date.end_date ||
            !isAfter(queryEndDate, startOfDay(new Date(date.end_date)))
          ) {
            // Construct response and stop looking for available dates
            response.push({
              id: studio.id,
              city: studio.city,
              state: studio.state,
              country: studio.country,
              studioName: studio.studio_name,
              availableStage: {
                stageId: stage.stage_id,
                stageNumber: stage.stage_number,
              },
            });
            break;
          }
        }
      }
    });
  });

  return response;
};
