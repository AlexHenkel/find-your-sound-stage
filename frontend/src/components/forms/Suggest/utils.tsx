import { ItemListPredicate } from '@blueprintjs/select';
import { Option } from './types';

function escapeRegExpChars(text: string) {
  return text.replace(/([.*+?^=!:${}()|[\]/\\])/g, '\\$1');
}

// Default `highglightText` function provdied by Blueprint docs
// https://github.com/palantir/blueprint/blob/develop/packages/docs-app/src/common/films.tsx
export function highlightText(text: string, query: string) {
  let lastIndex = 0;
  const words = query
    .split(/\s+/)
    .filter(word => word.length > 0)
    .map(escapeRegExpChars);
  if (words.length === 0) {
    return [text];
  }
  const regexp = new RegExp(words.join('|'), 'gi');
  const tokens: React.ReactNode[] = [];

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const match = regexp.exec(text);
    if (!match) {
      break;
    }
    const length = match[0].length;
    const before = text.slice(lastIndex, regexp.lastIndex - length);
    if (before.length > 0) {
      tokens.push(before);
    }
    lastIndex = regexp.lastIndex;
    tokens.push(<strong key={lastIndex}>{match[0]}</strong>);
  }
  const rest = text.slice(lastIndex);
  if (rest.length > 0) {
    tokens.push(rest);
  }
  return tokens;
}

/**
 * Returns the list of suggested items after a query input by the user
 *
 * Our select implementation supports suggesting items either from the label or value;
 * if we found an exact match, we should just return that item. Otherwise, return approximated items.
 * All comparisons are normalized to lower case to extend results
 */
export const getFilteredList: ItemListPredicate<Option> = (
  currQuery,
  items,
) => {
  const normalizedQuery = currQuery.toLowerCase();

  const exactMatch = items.find(
    item =>
      item.label.toLowerCase() === normalizedQuery ||
      item.value.toLowerCase() === normalizedQuery,
  );
  if (exactMatch) {
    return [exactMatch];
  }

  return items.filter(
    item => getText(item).toLowerCase().indexOf(normalizedQuery) >= 0,
  );
};

export const getText = (item: Option) => `${item.label} | ${item.value}`;
