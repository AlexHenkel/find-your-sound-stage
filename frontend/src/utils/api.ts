export interface ApiResponse {
  search: StudioResult[];
}

export interface StudioResult {
  id: number;
  studioName: string;
  city: string;
  state: string;
  country: string;
  availableStage: SoundStage;
}

export interface SoundStage {
  stageId: number;
  stageNumber: string;
}
