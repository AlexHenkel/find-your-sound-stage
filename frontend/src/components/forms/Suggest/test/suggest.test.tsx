import { getFilteredList } from '../utils';
import statesData from 'data/states.json';

/**
 * We asume original <Suggest /> component is being tested by Blueprint
 */
describe('<Suggest />', () => {
  test('Query exact match of value returns correct list', () => {
    const result = getFilteredList(statesData[0].value, statesData);
    expect(result).toMatchSnapshot();
  });

  test('Query exact match of label returns correct list', () => {
    const result = getFilteredList(statesData[0].label, statesData);
    expect(result).toMatchSnapshot();
  });

  test('Query arbitrary value returns correct list', () => {
    const result = getFilteredList('am', statesData);
    expect(result).toMatchSnapshot();
  });
});
