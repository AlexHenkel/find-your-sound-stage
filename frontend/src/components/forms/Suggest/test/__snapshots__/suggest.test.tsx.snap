// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Suggest /> Query arbitrary value returns correct list 1`] = `
Array [
  Object {
    "label": "AL",
    "value": "Alabama",
  },
  Object {
    "label": "AS",
    "value": "American Samoa",
  },
  Object {
    "label": "GU",
    "value": "Guam",
  },
  Object {
    "label": "NH",
    "value": "New Hampshire",
  },
]
`;

exports[`<Suggest /> Query exact match of label returns correct list 1`] = `
Array [
  Object {
    "label": "AL",
    "value": "Alabama",
  },
]
`;

exports[`<Suggest /> Query exact match of value returns correct list 1`] = `
Array [
  Object {
    "label": "AL",
    "value": "Alabama",
  },
]
`;
